bfapi/service/user.py

```diff
@@ -0,0 +1,204 @@
+    if row['geoaxis_uid'] != geoaxis_uid:
+        err = DatabaseResultError("select_user", "geoaxis_uid")
+        db.print_diagnostics(err)
+        raise err
+    return User(
+        geoaxis_uid=row['geoaxis_uid'],
+        bf_api_key=row['api_key'],
+        user_name=row['user_name'],
+        created_on=row['created_on'],
+    )
+
+def _db_harmonize(inp_user: User) -> User:
+    log = logging.getLogger(__name__)
+    if inp_user is None:
+        raise Exception("_db_harmonize called on empty User object")
+    db_user = get_by_id(inp_user.geoaxis_uid)
+    if db_user is None:
+        if not inp_user.bf_api_key:
+            inp_user.bf_api_key = uuid.uuid4()
+            
+        conn = db.get_connection()
+        transaction = conn.begin()
+        try:
+            db.user.insert_user(
+                conn,
+                geoaxis_uid=inp_user.geoaxis_uid,
+                user_name=inp_user.user_name,
+                api_key=inp_user.bf_api_key,
+            )
+            transaction.commit()
+        except db.DatabaseError as err:
+            transaction.rollback()
+            log.error('Could not save user to database: %s', err)
+            db.print_diagnostics(err)
+            raise
+        finally:
+            conn.close()
+        return inp_user
+    else:
+        if inp_user.user_name != "":
+            db_user.user_name = inp_user.user_name
+        if inp_user.bf_api_key:
+            db_user.bf_api_key = inp_user.bf_api_key
+        conn = db.get_connection()
+        transaction = conn.begin()
+        try:
+            db.user.update_user(
+                conn,
+                geoaxis_uid=inp_user.geoaxis_uid,
+                user_name=inp_user.user_name,
+                api_key=inp_user.bf_api_key,
+            )
+            transaction.commit()
+        except db.DatabaseError as err:
+            transaction.rollback()
+            log.error('Could not update user %s in database: %s', inp_user.geoaxis_uid, err)
+            db.print_diagnostics(err)
+            raise
+        finally:
+            conn.close()
+        return db_user
+
+#
+# Errors
+#
+
+class Error(Exception):
+    def __init__(self, message: str):
+        super().__init__(message)
+
+class DatabaseResultError(Error):
+    def __init__(self, func_name: str, db_entry: str):
+        super().__init__("Database result error: {} returned entry with incorrect {}.".format(func_name, db_entry))
+
+class CouldNotReachError(Error):
+    def __init__(self, targ_addr: str):
+        super().__init__('{} is unreachable'.format(targ_addr))
+
+class UnauthorizedError(Error):
+    def __init__(self, targ_addr: str):
+        super().__init__('Not authorized to access {}'.format(targ_addr))
+
+class HttpCodeError(Error):
+    def __init__(self, status_code: int, targ_addr: str):
+        super().__init__('Http Code error (HTTP {}) at {}'.format(status_code, targ_addr))
+        self.status_code = status_code
+
+class ResponseError(Error):
```

Comment:
Can we rename this to `InvalidGeoaxisResponse` for clarity?  `ResponseError` in the context of a module called user is fairly ambiguous.